### PR TITLE
Improve `oe_get_keys()`

### LINKED ISSUE #196: Get unique values for a given set of keys
See discussion in https://github.com/ropensci/osmextract/discussions/188

### LINKED ISSUE #201: Bug in oe_get_keys
Reprex: 

``` r
# packages
library(osmextract)
#> Data (c) OpenStreetMap contributors, ODbL 1.0. https://www.openstreetmap.org/copyright.
#> Check the package website, https://docs.ropensci.org/osmextract/, for more details.

# read in test data
iow <- oe_get("Isle of Wight", quiet = TRUE, query = "SELECT * FROM lines WHERE osm_id == 594620")

# Print the other_tags column
iow$other_tags
#> [1] "\"comment\"=>\"Removed junction=roundabout as some entrances have right of way. See http://wiki.openstreetmap.org/wiki/Tag:junction%3Droundabout\",\"lit\"=>\"yes\",\"maxspeed\"=>\"30 mph\",\"oneway\"=>\"yes\",\"surface\"=>\"asphalt\""
```

The keys should be “comment”, “lit”, “maxspeed”, “oneway” and “surface”, while

``` r
oe_get_keys(iow)
#> [1] "comment"                                                                                                         
#> [2] "roundabout as some entrances have right of way. See http://wiki.openstreetmap.org/wiki/Tag:junction%3Droundabout"
#> [3] "yes"                                                                                                             
#> [4] "30 mph"                                                                                                          
#> [5] "asphalt"
```

<sup>Created on 2021-05-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### PR DESCRIPTION
**Pull request by agila5**
Fix #201 and will fix #196. I will add a more detailed description later on.

### CODE REVIEW

**Comment by agila5** [comment]
Hi @Robinlovelace! Tomorrow I will add tests, fix docs, tidy code and so on. Meanwhile, this is the new approach: 

``` r
# packages
library(osmextract)
#> Data (c) OpenStreetMap contributors, ODbL 1.0. https://www.openstreetmap.org/copyright.
#> Check the package website, https://docs.ropensci.org/osmextract/, for more details.

# read in data
iow <- oe_get("Isle of Wight", quiet = TRUE)

# keys
head(oe_get_keys(iow), 10)
#>  [1] "lit"           "lanes"         "surface"       "maxspeed"     
#>  [5] "ref"           "sidewalk"      "cycleway:both" "incline"      
#>  [9] "oneway"        "prow_ref"

# keys and values
oe_get_keys(iow, values = TRUE)
#> access = {#private = 6523; #destination = 307; #permissive = 158; #customers...}
#> surface = {#asphalt = 3914; #unpaved = 643; #dirt = 421; #grass = 412; #grav...}
#> service = {#driveway = 3393; #parking_aisle = 758; #alley = 66; #siding = 23...}
#> lit = {#yes = 2728; #no = 1514}
#> designation = {#public_footpath = 2355; #public_bridleway = 986; #byway_open...}
#> prow_ref = {#BB10 = 18; #N207 = 18; #S3 = 16; #GL44 = 15; #SS2 = 15; #SW56 =...}
#> foot = {#designated = 1496; #yes = 1129; #permissive = 383; #destination = 5...}
#> maxspeed = {#30 mph = 2002; #60 mph = 442; #40 mph = 260; #20 mph = 156; #10...}
#> bicycle = {#yes = 1091; #no = 679; #designated = 211; #dismount = 147; #perm...}
#> horse = {#designated = 722; #no = 524; #yes = 326; #permissive = 40; #privat...}
#> [Truncated output...]

# just some keys
oe_get_keys(iow, values = TRUE, which_keys = c("access", "surface"))
#> access = {#private = 6523; #destination = 307; #permissive = 158; #customers...}
#> surface = {#asphalt = 3914; #unpaved = 643; #dirt = 421; #grass = 412; #grav...}
```

<sup>Created on 2021-05-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

If you have any suggestion please let me know. Moreover, if you test the code and you find any bug, please let me know.

**Comment by Robinlovelace** [comment]
Looking good Andrea, really like how they print.

**Comment by Robinlovelace** [comment]
Just tested for West Yorkshire and it's picked up an issue: 

``` r
remotes::install_github("ropensci/osmextract", ref = "key-values")
#> Using github PAT from envvar GITHUB_PAT
#> Skipping install of 'osmextract' from a github remote, the SHA1 (5935af79) has not changed since last install.
#>   Use `force = TRUE` to force installation
library(osmextract)
#> Data (c) OpenStreetMap contributors, ODbL 1.0. https://www.openstreetmap.org/copyright.
#> Check the package website, https://docs.ropensci.org/osmextract/, for more details.
iow = oe_get("Isle of Wight", quiet = TRUE)

head(oe_get_keys(iow), 10)
#>  [1] "lit"             "surface"         "sidewalk"        "incline"        
#>  [5] "designation"     "prow_ref"        "foot"            "horse"          
#>  [9] "not:name"        "source:maxspeed"

# bigger area
wy = oe_get("west yorkshire", quiet = TRUE)
wy_keys = oe_get_keys(wy)
length(wy_keys)
#> [1] 634
head(wy_keys, 10)
#>  [1] "lit"           "addr:postcode" "lcn"           "maxspeed"     
#>  [5] "sidewalk"      "oneway"        "lanes"         "ref"          
#>  [9] "gritting"      "maxweight"
# keys and values
wy_kvs = oe_get_keys(wy, values = TRUE)
#> Error: There are more keys than values (or vice versa). Please raise a new issue at https://github.com/ropensci/osmextract
```

<sup>Created on 2021-05-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by Robinlovelace** [comment]
This is interesting: 

```
> iow_kvs$tactile_paving
[1] "yes"
> sapply(iow_kvs, length)[1:9]
    surface     service         lit designation    prow_ref        foot       horse     natural   tracktype 
       5913        4046        3570        3425        3406        2975        1532        1499        1476 
```

I like how the keys with most results go first, this is great. Worth adding something like that to the docs also?

**Comment by agila5** [comment]
> Just tested for West Yorkshire and it's picked up an issue:

Thanks, will check tomorrow! 

> I like how the keys with most results go first, this is great. Worth adding something like that to the docs also?

At the end I will update all docs, vignettes and so on!

**Comment by agila5** [comment]
It seems that everything is working now: 

``` r
# packages
library(osmextract)
#> Data (c) OpenStreetMap contributors, ODbL 1.0. https://www.openstreetmap.org/copyright.
#> Check the package website, https://docs.ropensci.org/osmextract/, for more details.

# test west Yorkshire
wy <- oe_get("West Yorkshire", quiet = TRUE)

# keys and values
oe_get_keys(wy, values = TRUE)
#> surface = {#asphalt = 11117; #paved = 2760; #unpaved = 1564; #ground = 821;  ...}
#> maxspeed = {#30 mph = 7489; #20 mph = 5802; #40 mph = 2368; #60 mph = 1021;  ...}
#> service = {#parking_aisle = 9107; #driveway = 4099; #alley = 1371; #yard = 2 ...}
#> oneway = {#yes = 11920; #no = 1793; #-1 = 76; #alternating = 30; #reversible = 12}
#> foot = {#yes = 7188; #designated = 2896; #permissive = 494; #no = 381; #unkn ...}
#> access = {#private = 7058; #no = 999; #yes = 848; #customers = 583; #permiss ...}
#> lit = {#yes = 8756; #no = 1668; #sunset-20:00;05:30-sunrise = 18; #limited = ...}
#> ref = {#A629 = 377; #A61 = 331; #A58 = 327; #A62 = 294; #A638 = 284; #A65 =  ...}
#> source:name = {#OS_OpenData_Locator = 7345; #OS_OpenData_StreetView = 873; # ...}
#> lanes = {#2 = 4699; #1 = 1908; #3 = 1093; #4 = 143; #5 = 21; #6 = 4; #3;4;4  ...}
#> [Truncated output...]

# test greater london
gl <- oe_get("Greater London", quiet = TRUE)

# keys and values
oe_get_keys(gl, values = TRUE)
#> maxspeed = {#20 mph = 60209; #30 mph = 23177; #40 mph = 3221; #50 mph = 1595 ...}
#> lit = {#yes = 75119; #no = 5724; #24/7 = 125; #disused = 10; #automatic = 9; ...}
#> surface = {#asphalt = 49533; #paved = 6941; #paving_stones = 4999; #concrete ...}
#> oneway = {#yes = 40360; #no = 3978; #-1 = 7}
#> sidewalk = {#both = 17228; #left = 6819; #separate = 3550; #none = 1434; #ri ...}
#> ref = {#A406 = 534; #A4 = 514; #A3 = 412; #A12 = 370; #A10 = 365; #A40 = 360 ...}
#> service = {#parking_aisle = 11855; #driveway = 8909; #alley = 2690; #siding  ...}
#> access = {#private = 17659; #no = 2390; #permissive = 1315; #customers = 119 ...}
#> foot = {#yes = 15241; #designated = 3798; #no = 2547; #permissive = 2036; #p ...}
#> layer = {#1 = 11555; #-1 = 4960; #-2 = 893; #2 = 792; #-3 = 424; #0 = 308; # ...}
#> [Truncated output...]

# test nord-ovest italy
no <- oe_get("Nord-ovest", quiet = TRUE)

# keys and values
oe_get_keys(no, values = TRUE)
#> surface = {#asphalt = 244703; #ground = 49372; #unpaved = 19035; #gravel = 1 ...}
#> oneway = {#yes = 193912; #no = 54575; #alternating = 155; #reversible = 88;  ...}
#> layer = {#-1 = 67055; #1 = 55990; #0 = 1476; #-2 = 1309; #2 = 1092; #3 = 253 ...}
#> service = {#parking_aisle = 54226; #driveway = 52357; #yard = 5693; #alley = ...}
#> tracktype = {#grade3 = 35132; #grade2 = 33999; #grade4 = 21693; #grade5 = 16 ...}
#> ref = {#SS1 = 1089; #SP2 = 882; #SP10 = 860; #SP1 = 721; #A10 = 711; #SP3 =  ...}
#> lanes = {#2 = 62188; #1 = 39632; #3 = 3924; #4 = 1149; #5 = 126; #6 = 45; #7 ...}
#> access = {#private = 71474; #yes = 8204; #no = 6608; #agricultural = 5246; # ...}
#> lit = {#yes = 84610; #no = 17314; #24/7 = 677; #dusk-dawn = 53; #automatic = ...}
#> maxspeed = {#50 = 46377; #30 = 17121; #40 = 9850; #90 = 5376; #70 = 5245; #1 ...}
#> [Truncated output...]
```

<sup>Created on 2021-05-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

I will update the docs and tests, and then it's ready for review!

**Comment by agila5** [comment]
Summary of the new approach: 

``` r
# packages
library(osmextract)
#> Data (c) OpenStreetMap contributors, ODbL 1.0. https://www.openstreetmap.org/copyright.
#> Check the package website, https://docs.ropensci.org/osmextract/, for more details.

its_path = oe_get("ITS Leeds", download_only = TRUE)
#> The input place was matched with: ITS Leeds
#> The chosen file was already detected in the download directory. Skip downloading.
#> The corresponding gpkg file was already detected. Skip vectortranslate operations.

# Get keys from an OSM extract
oe_get_keys("ITS Leeds")
#>  [1] "surface"             "lanes"               "bicycle"            
#>  [4] "lit"                 "access"              "maxspeed"           
#>  [7] "ref"                 "foot"                "natural"            
#> [10] "lanes:backward"      "lanes:forward"       "source:name"        
#> [13] "step_count"          "lanes:psv:backward"  "alt_name"           
#> [16] "layer"               "motor_vehicle"       "tunnel"             
#> [19] "bridge"              "covered"             "incline"            
#> [22] "lanes:psv"           "service"             "turn:lanes"         
#> [25] "turn:lanes:forward"  "frequency"           "indoor"             
#> [28] "lcn"                 "level"               "maxheight"          
#> [31] "operator"            "power"               "source:geometry"    
#> [34] "substation"          "turn:lanes:backward" "voltage"            
#> [37] "website"

# Get keys and values
oe_get_keys("ITS Leeds", values = TRUE)
#> surface = {#asphalt = 12; #paved = 3; #cobblestone = 1; #paving_stones = 1}
#> lanes = {#2 = 9; #1 = 7}
#> bicycle = {#yes = 10; #designated = 5}
#> lit = {#yes = 15}
#> access = {#permissive = 12; #yes = 2}
#> maxspeed = {#30 mph = 12}
#> ref = {#A660 = 9; #4184 = 1}
#> foot = {#yes = 5; #designated = 4}
#> natural = {#tree_row = 7}
#> lanes:backward = {#2 = 3; #1 = 1; #3 = 1}
#> [Truncated output...]

# Subset some keys
oe_get_keys("ITS Leeds", values = TRUE, which_keys = c("surface", "lanes"))
#> surface = {#asphalt = 12; #paved = 3; #cobblestone = 1; #paving_stones = 1}
#> lanes = {#2 = 9; #1 = 7}

# Print all (non-NA) values for a given key
oe_get_keys("ITS Leeds", values = TRUE)["surface"]
#> $surface
#>  [1] "asphalt"       "asphalt"       "asphalt"       "asphalt"      
#>  [5] "asphalt"       "asphalt"       "paved"         "cobblestone"  
#>  [9] "asphalt"       "asphalt"       "paved"         "paved"        
#> [13] "paving_stones" "asphalt"       "asphalt"       "asphalt"      
#> [17] "asphalt"

# Get keys from an existing sf object
its = oe_get("ITS Leeds")
#> The input place was matched with: ITS Leeds
#> The chosen file was already detected in the download directory. Skip downloading.
#> The corresponding gpkg file was already detected. Skip vectortranslate operations.
#> Reading layer `lines' from data source `C:\Users\Utente\Documents\osm-data\test_its-example.gpkg' using driver `GPKG'
#> Simple feature collection with 189 features and 10 fields
#> Geometry type: LINESTRING
#> Dimension:     XY
#> Bounding box:  xmin: -1.562458 ymin: 53.80471 xmax: -1.548076 ymax: 53.81105
#> Geodetic CRS:  WGS 84
oe_get_keys(its, values = TRUE)
#> surface = {#asphalt = 12; #paved = 3; #cobblestone = 1; #paving_stones = 1}
#> lanes = {#2 = 9; #1 = 7}
#> bicycle = {#yes = 10; #designated = 5}
#> lit = {#yes = 15}
#> access = {#permissive = 12; #yes = 2}
#> maxspeed = {#30 mph = 12}
#> ref = {#A660 = 9; #4184 = 1}
#> foot = {#yes = 5; #designated = 4}
#> natural = {#tree_row = 7}
#> lanes:backward = {#2 = 3; #1 = 1; #3 = 1}
#> [Truncated output...]

# Get keys from a character vector pointing to a file (might be faster than
# reading the complete file)
oe_get_keys(its_path, values = TRUE)
#> surface = {#asphalt = 12; #paved = 3; #cobblestone = 1; #paving_stones = 1}
#> lanes = {#2 = 9; #1 = 7}
#> bicycle = {#yes = 10; #designated = 5}
#> lit = {#yes = 15}
#> access = {#permissive = 12; #yes = 2}
#> maxspeed = {#30 mph = 12}
#> ref = {#A660 = 9; #4184 = 1}
#> foot = {#yes = 5; #designated = 4}
#> natural = {#tree_row = 7}
#> lanes:backward = {#2 = 3; #1 = 1; #3 = 1}
#> [Truncated output...]

# Add an extra key to an existing .gpkg file without repeating the
# vectortranslate operations
colnames(its)
#>  [1] "osm_id"     "name"       "highway"    "waterway"   "aerialway" 
#>  [6] "barrier"    "man_made"   "oneway"     "z_order"    "other_tags"
#> [11] "geometry"
colnames(oe_read(
  its_path,
  query = "SELECT *,  hstore_get_value(other_tags, 'oneway') AS oneway FROM lines",
  quiet = TRUE
))
#>  [1] "osm_id"     "name"       "highway"    "waterway"   "aerialway" 
#>  [6] "barrier"    "man_made"   "oneway"     "z_order"    "other_tags"
#> [11] "geometry"
```

<sup>Created on 2021-05-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by Robinlovelace** [comment]
Outstanding work Andrea, very happy with this approach, looking good and ready to merge!

One thought, on reading this:

```
oe_get_keys(wy, values = TRUE)
#> surface = {#asphalt = 11117; #paved = 2760; #unpaved = 1564; #ground = 821;  ...}
...
```

Worth adding a line before that saying how many unique keys have been found, e.g.:

```
oe_get_keys(wy, values = TRUE)
#>  900 unique keys found the first 10 of which, in ascending order of % NA values, is:
#> surface (10% NAs) = {#asphalt = 11117; #paved = 2760; #unpaved = 1564; #ground = 821;  ...}
```

Note also I think it could be good to say what the % NAs is for each key seeing is that is the most common value. Overall these are minor comments and I think this is good to go.

File: tests/testthat/test-get-keys.R [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  expect_error(
+    oe_get_keys(sf::st_sfc(sf::st_point(c(1, 1)), crs = 4326)),
+    "there is no support for objects of class"
+  )
+  expect_error(oe_get_keys("xxx.gpkg")) # file does not exist
+  expect_error(oe_get_keys(c("a.gpkg", "b.gpkg"))) # length > 1
+})
+
+test_that("oe_get_keys: reads from sf object", {
+  its_object <- oe_read(its_pbf, skip_vectortranslate = TRUE, quiet = TRUE)
```
Comment (Robinlovelace) [inline tests/testthat/test-get-keys.R:86]: Please use `=` assignment. See https://github.com/ropensci/software-review-meta/issues/71#issuecomment-445304208

File: tests/testthat/_snaps/get-keys.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# oe_get_keys + values: printing method
```
Comment (Robinlovelace) [inline tests/testthat/_snaps/get-keys.md:1]: Very cool, good to see use of snaps, a capability of testthat 3 I think?!

**Review by Robinlovelace — APPROVED** [review]
I have a minor comment about assignment but overall this is a huge improvment - next step JSS paper :rocket:

File: tests/testthat/_snaps/get-keys.md (hunk 2 above)
Comment (agila5) [reply tests/testthat/_snaps/get-keys.md:1]: Yup! See also the DESCRIPTION file

**Comment by agila5** [comment]
New output: 

``` r
library(osmextract)
#> Data (c) OpenStreetMap contributors, ODbL 1.0. https://www.openstreetmap.org/copyright.
#> Check the package website, https://docs.ropensci.org/osmextract/, for more details.
oe_get_keys("West Yorkshire", values = TRUE)
#> Found 723 unique keys, printed in ascending order of % NA values. The first 10 keys are: 
#> surface (90% NAs) = {#asphalt = 11117; #paved = 2760; #unpaved = 1564; #groun...}
#> maxspeed (91% NAs) = {#30 mph = 7489; #20 mph = 5802; #40 mph = 2368; #60 mph...}
#> service (93% NAs) = {#parking_aisle = 9107; #driveway = 4099; #alley = 1371; ...}
#> oneway (94% NAs) = {#yes = 11920; #no = 1793; #-1 = 76; #alternating = 30; #r...}
#> foot (95% NAs) = {#yes = 7188; #designated = 2896; #permissive = 494; #no = 3...}
#> access (95% NAs) = {#private = 7058; #no = 999; #yes = 848; #customers = 583;...}
#> lit (95% NAs) = {#yes = 8756; #no = 1668; #sunset-20:00;05:30-sunrise = 18; #...}
#> ref (95% NAs) = {#A629 = 377; #A61 = 331; #A58 = 327; #A62 = 294; #A638 = 284...}
#> source:name (96% NAs) = {#OS_OpenData_Locator = 7345; #OS_OpenData_StreetView...}
#> lanes (96% NAs) = {#2 = 4699; #1 = 1908; #3 = 1093; #4 = 143; #5 = 21; #6 = 4...}
#> [Truncated output...]
```

<sup>Created on 2021-05-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Pull request merged**